elytron/WFLY-16306-keycloak-saml-feature-pack.adoc

```diff
@@ -0,0 +1,331 @@
+
+A new Galleon feature pack will be added to the Keycloak project with the group ID, `org.keycloak`, and artifact ID, `keycloak-saml-adapter-galleon-pack`.
+This new feature pack will contain three layers called `keycloak-saml`, `keycloak-client-saml`, and `keycloak-client-saml-ejb`.
+
+When this feature pack and its layers are provisioned to a WildFly installation, the end result will mostly match the results
+of manually installing the Keycloak Elytron SAML adapter using the instructions linked above.
+
+In particular, when the `keycloak-saml` layer is provisioned, the following steps will occur:
+
+ # Install a set of modules under `modules/system/add-ons`
+ # Register the `org.keycloak.keycloak-saml-adapter-subsystem` extension.
+ # Add an empty `urn:jboss:domain:keycloak:1.4` subsystem.
+
+The `keycloak-client-saml` layer will depend on the `keycloak-saml` layer and when provisioned, the following additional steps will occur:
+
+ # Add the required resources to the `elytron` subsystem to result in an `http-authentication-factory` and `security-domain` supporting SAML being available.
+ # Update the `application-security-domain` mapping in the `undertow` subsystem to map the default security domain to the new `http-authentication-factory`.
+
+The `keycloak-client-saml-ejb` layer will depend on the `keycloak-client-saml` layer and when provisioned, the following additional steps will occur:
+
+ # Update the `application-security-domain` mapping in the `ejb3` subsystem to map the default security domain to the new `KeycloakDomain`.
+
+Note that the existing CLI scripts for the SAML adapter and the existing feature pack for OIDC only update the `application-security-domain` mapping in the
+`undertow` subsystem. However, the CLI scripts for the OIDC adapter also update the `application-security-domain`
+mapping in the EJB3 subsystem. For this reason, we have added the `keycloak-client-saml-ejb` layer here for users
+that need to propagate their identities to EJBs as well.
+
+The CLI scripts that would have been added via the manual installation steps won't be added to the resulting installation
+since the required resources will be added already during provisioning.
+
+Note that Keycloak and WildFly have different release cadences. This means anyone using Galleon to keep their WildFly
+installation up to date should be able to receive updates to the core WildFly installation independently of the new
+Keycloak feature pack. Having the new Keycloak feature pack installed should not prevent the server from being updated
+once later WildFly versions are available. Similarly, WildFly releases should not trigger the need for an update to Keycloak.
+
+==== Bare Metal vs. Cloud Use Cases
+
+For the bare metal use case, the `keycloak-client-saml` layer should be used.
+
+For WildFly s2i, when using the automatic registration of SAML deployments feature, the `keycloak-saml` layer should be used.
```

Comment:
@fjuma @jfdenise please help me with the following: where can I find some explanation for automatic registration of SAML deployments VS  manual deployments? I don't know how they work (or perhaps I can't find a correspondence with things I know) and hence, this part is not clear to me...

Reply:
since different layers are used, I suppose different deployments are needed... is that correct?

Reply:
If I get it right. I should use `keycloak-saml` instead of `keycloak-client-saml` layer to generate CLI configuration at boot time. The necessary information for configuration are provided in SSO_XXX variables. It was specified for https://issues.redhat.com/browse/EAP7-1947. This is automatic registration of SAML deployments.

For manual registration of SAML deployments. I should use `keycloak-client-saml` and add CLI script manualy.

I am right @fjuma @jfdenise? Please add some explanation if I am wrong.

Reply:
For the bare metal use case, the `keycloak-client-saml` layer should be used.

For WildFly s2i, when using the automatic registration of SAML deployments feature, the `keycloak-saml` layer should be used instead because it handles adding the required resources to the `elytron` subsystem configuration.

Reply:
@fjuma I see that layer `keycloak-client-saml` adds the following extra config (some omitted) if compared to layer `keycloak-saml`:

```
<http-authentication-factory name="keycloak-http-authentication" security-domain="KeycloakDomain" http-server-mechanism-factory="keycloak-http-server-mechanism-factory">
                    <mechanism-configuration>
                        <mechanism mechanism-name="KEYCLOAK-SAML">
                            <mechanism-realm realm-name="KeycloakSAMLRealm" realm-mapper="keycloak-saml-realm-mapper"/>
                        </mechanism>
                    </mechanism-configuration>
                </http-authentication-factory>
...

            <application-security-domains>
                <application-security-domain name="other" http-authentication-factory="keycloak-http-authentication"/>
            </application-security-domains>
```

looks like that, when using `keycloak-client-saml`, all applications are secured by `KEYCLOAK-SAML` by default: is that correct?

Also, I don't see any reference to the RHSSO server: can you confirm https://github.com/wildfly/wildfly-proposals/pull/486#discussion_r1086713579 which basically is "Use `keycloak-saml` + `SSO_XXX` for s2i; use `keycloak-client-saml` + cli for baremetal & bootable jar"

Reply:
@tommaso-borgato The same config will also be added automatically by a CLI script when using WildFly s2i with the automatic registration.

This configuration doesn't mean that all applications are secured by KEYCLOAK-SAML by default. To secure an app with KEYCLOAK-SAML, either `<auth-method>KEYCLOAK-SAML</auth-method>` needs to be specified in the app's `web.xml` file or a `secure-deployment` that references the runtime name of the application needs to be configured in the `keycloak-saml` subsystem.

The SSO_XXX variables would be specified when deploying an application using the Helm chart.

@jfdenise Is there some documentation about the SSO_XXX variables for SAML that you could point @tommaso-borgato  to?

Reply:
@tommaso-borgato , the list of env variables can be found there: https://github.com/wildfly/wildfly-cekit-modules/blob/main/jboss/container/wildfly/launch/keycloak/2.0/module.yaml

Reply:
@fjuma @jfdenise I need to understand y what exactl `WildFly s2i using the automatic registration of SAML deployments feature` and `WildFly s2i with manual deployments` mean ... 

IIUC, the extra configuration added by the `keycloak-client-saml` layer, when compared to `keycloak-saml` layer, has nothing to do with the deployment (the war file):

```
- Add the required resources to the `elytron` subsystem to result in an `http-authentication-factory` and `security-domain` supporting SAML being available.
- Update the `application-security-domain` mapping in the `undertow` subsystem to map the default security domain to the new `http-authentication-factory`.
```

Reply:
The SAML automatic registration does the following:
* Configure elytron, undertow and ejb security (so simlira to what the keycloak-client-saml).
* Discover the war to be secured with saml and update the keycloak-saml subsystem.
* Handle some automatic registration of the client in the keycloak server.

Reply:
and do this correspond to usual `s2i` workflow using the `wildfly-maven-plugin`? if the answer is yes, is there an example on `WildFly s2i using the automatic registration of SAML deployments feature` and one on `WildFly s2i with manual deployments` to help me catch the differences?

Reply:
@jfdenise how is `SAML automatic registration` triggered? I mean how can I use `SAML automatic registration` or `SAML manual registration`? is there any example that might help me?

Reply:
SAML automatic registration is triggered when you set SSO_XXX env variables. The manual registration, you have to install generated CLI scripts generated from the Keycloak console in your application src code, these would get copied to the server installation and discovered at startup. You can look at this testcase: https://github.com/wildfly/wildfly-s2i/tree/wf-26.0/test/test-app-keycloak/configuration

Reply:
reporting here explanation by @fjuma in order to answer this thread: 

```
The manual registration means that either the deployment needs to contain a keycloak-saml.xml file or a user provided script that configures the keycloak-saml subsystem needs to be provided

The benefit of the automatic registration when using is s2i is that it takes care of creating the Keycloak client that's needed and it automatically creates the script to configure the keycloak-saml subsystem
```

The user provided script can be obtained once the client has been configured in keycloak and is the one containing e.g.:

```
/subsystem=keycloak-saml/secure-deployment=simple-webapp.war/:add
...
```